maven-release-manager/src/main/java/org/apache/maven/shared/release/config/ReleaseDescriptorBuilder.java

```diff
@@ -58,7 +59,9 @@ public ReleaseDescriptorBuilder addCheckModificationExclude( String string )
 
     public ReleaseDescriptorBuilder setActivateProfiles( List<String> profiles )
     {
-        releaseDescriptor.setActivateProfiles( profiles );
+        List<String> copy = new ArrayList<>();
+        copy.addAll( profiles );
```

Comment:
This now acts as a merge, not as a setter, so this can't be the right approach.

Reply:
This is not a merge, I do not keel the old value. It is a defensive setter, ensuring that the list implementation used is not read only / immuable

Reply:
we will test ASAP

Reply:
@bguerin I confirm that the fix works.
I would also suggest to change the test case [here](https://github.com/apache/maven-release/blob/3efc9993aae247cff8e2c7dcd97113a7ef3e69d0/maven-release-manager/src/test/java/org/apache/maven/shared/release/DefaultReleaseManagerTest.java#L728) and use the same approach at [few lines above](https://github.com/apache/maven-release/blob/3efc9993aae247cff8e2c7dcd97113a7ef3e69d0/maven-release-manager/src/test/java/org/apache/maven/shared/release/DefaultReleaseManagerTest.java#L721).
In short replace:
`secondBuilder.setActivateProfiles( new ArrayList( Arrays.asList("aProfile", "bProfile") ) );`
with
`secondBuilder.setActivateProfiles( Arrays.asList( Arrays.asList("aProfile", "bProfile") ) );`

Reply:
>  I confirm that the fix works.

YOUPI !!!

> change the test case

Ok, no problem, I just changed the test I wrote thinking I should not change others

Reply:
I still don't like this approach and think any value of the releaseDescriptor should be immutable. With the introduction of the release-api it should be easier to write your own phases, but I don't want any of the phases to change the release descriptor.

Reply:
If I understand right here you would to instantiate a new ReleaseDescriptorBuilder with the sum of active profiles something like:
```
Properties props = new Properties();
props.put("exec.activateProfiles", profiles);
ReleaseUtil.copyPropertiesToReleaseDescriptor(props , releaseDescriptor)
```

Reply:
Looks a bit weird, right? I can imagine there should be an extra method or class. 
Based on this I discovered that ReleaseUtils should be adjusted a bit:
```
    public static ReleaseDescriptor buildReleaseDescriptor( ReleaseDescriptorBuilder builder )
    {
        return builder.build();
    }
```

It exposes another hack: `config.setCompletedPhase( name );`, but this one is easy to fix.

Reply:
I did get. What should do `config.setCompletedPhase( name )` ?

Reply:
I had a look to the history of this file and also to the previous PR. I have to agree with @rfscholte . The behavior to change the list o unmodifiable here is not clean and does something different than other methods that also take a list.
Looking the deep the #50 it's better rework how the active profiles are [added to the ReleaseDescriptor](https://github.com/apache/maven-release/blob/b968279f72a12daea1836b795125a803bf7e8ccd/maven-release-manager/src/main/java/org/apache/maven/shared/release/DefaultReleaseManager.java#L317)

I do not understand why the [performRequest.getReleaseDescriptorBuilder()](https://github.com/apache/maven-release/blob/b968279f72a12daea1836b795125a803bf7e8ccd/maven-release-manager/src/main/java/org/apache/maven/shared/release/DefaultReleaseManager.java#L304) has a list of active profile and than after loadReleaseDescriptor [this](https://github.com/apache/maven-release/blob/b968279f72a12daea1836b795125a803bf7e8ccd/maven-release-manager/src/main/java/org/apache/maven/shared/release/DefaultReleaseManager.java#L307) some profile has gone so that is needed to be add them again.

Reply:
> ReleaseUtils during property copy overrides any existing value of the builder. The builder does not have getter...so we can not merge neither in ReleaseUtil neither in DefaultReleaseManager so what should be done

Yep, this is it

> ReleaseUtil must have the knowlegde that profiles are gathered from many parts so it should handle the copy like a merge

I can't find history of my previous PR anymore, but if I am right, this is what I start to do, and Robert said he prefered to see this specific code in DefaultReleaseManager ...

Reply:
`ReleaseUtil` should stay as stupid as possible. `copyPropertiesToReleaseDescriptor` means: if property exists, apply that to release descriptor.

Reply:
In the `DefaultReleaseManager` you can already see some anonymous classes that are used to control such logic.

Reply:
> copyPropertiesToReleaseDescriptor means: if property exists, apply that to release descriptor
apply == set (override) or merge (keep existing if possible) ?

Reply:
I had the same question

Reply:
there are 3 ways of merging when having a lefthandside and a righthandside:
- LHS over RHS ( setter )
- RHS over LHS ( setter )
- In case of Collections: combining LHS with RHS ( appender )

Reply:
> 
> 
> In the `DefaultReleaseManager` you can already see some anonymous classes that are used to control such logic.

I see. Maybe could be the case implement an inner class MergeReleaseDescriptorBuilder and replace all the instancies unless in some cases you really want to merge only a specific properties. Anyway this way it's quite tricky, I say tricky because it's applied more times so the idea of a ReleaseDescriptorBuilder unmutable it's not so strong.
```
ReleaseDescriptor releaseDescriptor =
            loadReleaseDescriptor( new ReleaseDescriptorBuilder() {
                       public void setActivateProfiles( List<String> profiles ) {
                             /* merge with */ performRequest.getReleaseDescriptorBuilder(),
                       }
             }, performRequest.getReleaseManagerListener() );
```

Reply:
> there are 3 ways of merging ..

You do not answer my question. When you say

> copyPropertiesToReleaseDescriptor means: if property exists, apply that to release descriptor

What should it do in case of a collection ? what "apply" means in your sentence : override existing value, or append to it ?

Reply:
`copyPropertiesToReleaseDescriptor` means `set`